src/engraving/compat/mscxcompat.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "mscxcompat.h"
+
+#include <QFile>
+#include <QBuffer>
+
+#include "log.h"
+
+Ms::Score::FileError mu::engraving::compat::loadMsczOrMscx(Ms::MasterScore* score, const QString& path, bool ignoreVersionError)
+{
+    QByteArray msczData;
+    QString filePath = path;
+    if (path.endsWith("mscx")) {
+        //! NOTE Convert mscx -> mscz
+
+        QFile mscxFile(path);
+        if (!mscxFile.open(QIODevice::ReadOnly)) {
+            LOGE() << "failed open file: " << path;
+            return Ms::Score::FileError::FILE_OPEN_ERROR;
+        }
+
+        QByteArray mscxData = mscxFile.readAll();
+
+        QBuffer buf(&msczData);
+        MsczWriter writer(&buf);
+        filePath = path + ".mscz";
+        writer.setFilePath(filePath);

```

Comment:
Is that assuming it's always OK to write to a file in the current directory that has the same name as the .mscx file but with .mscz appended to the end? Is that safe/sensible?
Is the a reason we only want Ms::Score( ) to be able to read from an .mscz file on disk, hence meaning we have to compress an mscx into a .mscz file first before it can then uncompress and parse it?

Reply:
We are planning to add new data storage to `mscz`, such as mixer settings, VST settings, etc.
We do not want to add this directly to the score xml file.
Moreover, we want to remove some data from score xml file, for example, styles, and store them separately.
Also, for example, we can separately store meta-information about the score - the author, title, etc., so as not to open and read the entire score to get it.
As a result, it turns out that the container will store many different files with different information, similar to how "open documents" are stored.
Accordingly, we will have to abandon the `mscx` format.
Therefore, I made it so that the main format is `mscz`, and `mscx` is opened in compatibility mode.

Reply:
Yes I gathered that (and I agree it makes sense), but when reading an mscx file, creating a dummy mscz file with a particular filename first just doesn't seem like the best approach when Score can already read an mscx directly.

Reply:
This file is created in memory (in a buffer, not in the filesystem).
I believe that we need to lighten `libmscore` as much as possible, it is already heavily overloaded with various code.
I think the opening of `mscx` will be only for tests, and very little is needed by users.

Reply:
Ah ok, the use of "setFilePath" made it look like it was writing to a file, that was really my only concern (though sure, compressing just to decompress again seems a bit inefficient, but unlikely to be a real problem).

Reply:
> Ah ok, the use of "setFilePath" made it look like it was writing to a file, that was really my only concern (though sure, compressing just to decompress again seems a bit inefficient, but unlikely to be a real problem).

Unfortunately, despite the fact that we are working with a buffer, we need to set the name of the files at the moment. Because the name of the score xml is determined by the name of the file.
I believe that we can change this and store the score xml with a constant name like `score.mscx`.